src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,9 +390,21 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                try
+                {
+                    int numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
+                    int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
```

Comment:
>Utils.GetIdentityPermutation(numBlocks) [](start = 55, length = 39)

Make `blockOrder` an `IEnumerable<int>`, and replace this specifiy "IdentityPermutation" with `Enumerable.Range(0, numBlocks)`. This avoids materializing the trivial case. I don't see a way to avoid materializing the less trivial case, I'm sorry to say. :(